### ISSUE TITLE
План работы по проекту cppTickets

### ISSUE THREAD

**Issue by m1nsk** [levelp/CppTickets#5]
- Описание Абстрактного класса Tickets и его методов.(Без описания абстрактного класса нет смысла браться за дочерние)
- Распределение работы по описанию дочерних классов между участниками
- Описание дочерних классов и методов в едином стиле.
- Реализация методов абстрактного и дочерних классов
- Описание и Реализация класса "Касса"
- Написание тестов для проверки работы классов.
- Долго рефакторим

В комментах жду предложения по параметрам и методам абстрактного класса Ticket. Потом выберем из них те, которые действительно являются существенными.

**Comment by alexey257** [comment]
Подумал, что-то как-то сложно получилось. Кроме билета надо класс кассы, источника билетов, и покупателя. В билете добавил поля состояния, каждый билет помнит свое состояние. В кассе автомат с switch-case, управляющий состоянием билета. Билеты можно положить массивом или списком, чтобы был произвольный доступ, очередь не пойдет. Источник кладет билет, покупатель бронирует, оплачивает, покупает. Касса переключает состояния билетов. Может быть сделать очередь покупателей. Может не делать классы источника и покупателя, а из main методы кассы стрелять.
Набросал кассу и какие в голову пришли поля билета, но ничего не компилил еще. Скажите, вообще по-другому сделать хочется или поправить-дополнить.

**Comment by m1nsk** [comment]
Норм, только, как я уже говорил, надо сначала разобраться с классом Ticket, потом дочерними, потом уже только браться за класс Касса. Например описанный статус билета-это свойство не кассы, а билета. Касса-некая база данных в которой собрана вся информация о билетах: статус, владелец, время, тип транспорта и т.д. В ней должны реализовываться паблик методы для покупателя, например заказ/бронь билета с такими-то параметрами и прайват типа изменения статуса билета, указание владельца билета. По сути это все что делает касса, работает с БД, в кассе не создаются новые билеты при приобретении, они все есть там изначально(ну если конечно в самолете не поставили пару новых кресел или к поезду не прицепили вагон, а в подлодке открыли форточку), покупатель лишь меняет их статус и указываем владельца. БД проще для нас организовать в виде таблицы(двумерного массива), тогда будет проще искать по id, статусу, типу транспорта. Но что-то я сам углубился в Кассу=) Короч, передлагаю такие параметры как время отправления, тип транспорта, место назначения,статус, id, владелец, цена, скидка.

**Comment by m1nsk** [comment]
На счет БД в виде таблицы, я наврал.
Описал Ticket класс, так что, если есть желающие пофантазировать на счет подводных лодок, то Велкам. Только все параметры в приват, доступ через методы, как у меня в Ticket

**Comment by JamesForster** [comment]
Думаю нужно для начала определиться с тем какие данные мы вводим и какие выводим, какие данные у нас исходные, в каком виде хранятся, тогда понятней будет что и как реализовывать.

**Comment by alexey257** [comment]
Билеты же постоянно поступают. Назначается рейс и появляются билеты c новыми номерами. Рейс заканчивается - билеты уничтожаются, тогда expired состояние не нужно. Состояния билета ступил, надо хранить в самом билете. Поля тоже в public зарядил зря. Думал БД сделать прямо в кассе, хотя конечно это будка, её можно опрокинуть:) Тогда БД либо настоящая либо класс. Несколько типов билетов по идее это несколько связанных таблиц по ID. Если класс, то там тоже несколько наследованных надо сделать и по get() из кассы нужный отдавать по указателю. Указатель наверное типа void и потом приведение к нужному типу, раз билеты разные, не знаю как это правильно делать.

А зачем переключатель снесли? его можно сразу в класс кассы утащить и потом дописать или вообще одновременно. Города тоже таблица в БД, можно добавить поля видов транспорта, cами города прямо текстом и ID. enum это злобно, он контроллерам больше нужен, состояния регистров читаемыми делать.

В общем, получается нужны классы билета с наследниками, БД, касса, а все действия из main. Полей в билете наверное хватит, надо наследников описать. 

Пока сочинял, коммент появился
Входные наверное создание билета от центра полетов или чего-то такого и запрос от покупателя, выходные - наверное отчет хозяину кассы сколько и чего продано.

**Comment by m1nsk** [comment]
Так-то правильно говоришь на счет БД. Но городить БД для городов тоже сурово, так мы рано или поздно напишем реально применимую прогу для заказа билетов=) Для городов достаточно словаря, пожалуй, хотя по-хорошему нужно учитывать в каких городах есть порт и куда летают самолеты. А что переключатели? Если Авейлбл, то статус авейлбл, если пейд, то пейд. Это как если писать Switch(a){
case(1): {
num=1;
}
}
то же самое можно num=a;
"
Пока сочинял, коммент появился
Входные наверное создание билета от центра полетов или чего-то такого и запрос от покупателя, выходные - наверное отчет хозяину кассы сколько и чего продано.
"
Это уже даже не кассы, это уже создание полноценных баз данных и классы работы с ними
Базы данных давайте пока сделаем статичными, по одной базе данных на каждый вид транспорта, поскольку классы у них разные. Предлагаю БД сделать тупо вектором, то есть массивом. Группа только вчера реализацию очереди осилила, а теперь базы данных пишет. Шикарно.

P.S. Код, кстати рабочий, можете покомпилить/запустить. Поменять set/get параметры. Но, там еще править кучу всего. Методы не проверяют вводимые диапазоны, как например время. Статус выводится в виде Int, а можно выводить строкой через switch/case, нет классов самолета/поезда/подлодки. Я сейчас создам Isuues на эти вещи, найдите задачу себе по силам и сделайте.
alexey257 давай накарябаем сегодня кассу?

**Comment by Lisss13** [comment]
У меня есть предложение, давайте с начало проработаем функцию main() в том плане, какие входные данные будут а заем относительно входных данных будем прорабатывать классы

**Comment by Lisss13** [comment]
Напишите кто какой класс будет делать

**Comment by m1nsk** [comment]
Зачем прорабатывать функцию main??? Это вообще последнее что будет прорабатываться. В main будут тесты и все. У нас задача прописать класс Ticket. В нем уже описывается все то, что можно сделать из мейн его методами. Если ты имеешь ввиду то, какие действия могут производиться с билетами от лица пользователя, то это реализуется не в мейн, а в классе касса. Ты когда покупаешь билет на самолет, не можешь изменить пункт назначения или цену на билет, или место в салоне, ты лишь привязываешь уже существующий билет к своим паспортным данным и меняешь его статус на оплачен, так что никто другой не может его купить.

**Comment by Lisss13** [comment]
Может составленный билет выгружать в файл, тем самым создавая базу данных?

**Comment by m1nsk** [comment]
Точно! А может еще Объектно-реляционное отображение(ORM) прикрутим? А потом реально начнем билеты на подлодки продавать? 
А может кто-то кроме Алексея хоть строчку напишет, это как бы на завтра задание? Я могу вот сейчас сесть и написать вот все к завтрашнему дню. Но это же вроде командная работа?

**Comment by alexey257** [comment]
Базу наверное проще, видел на других курсах - сделать классом. Реально когда Денис покажет БД, это будет думаю открытие интерфейса, потом create table, insert into и select from... where и больше мучений с тем, как правильно таблицы соединить, ключи задать.
В общем согласен, что переключатель-автомат можно здесь грохнуть и просто продавать как пирожки, поле у билета менять. Но в реальности всегда есть переходы состояния и действия в каждом case разные. Просто сейчас по работе это мне понятней и пока думал, понял, что там хоть какие поля у билета описывай, всё равно пока базовые действия-переходы не введем, даже не посмотреть, что что-то работает. Сейчас поля у билета есть, можно хоть на базовом классе кататься. Добавить билет, купить, вызвать статистику, что продали. 

Только увидел - m1nsk - про кассу. Хочу написать заготовку, но поспать бы и еще задачи с бумажки. Я за то, чтобы оживить в минимальном виде, и завтра на занятии решить кто чего писать будет.

**Comment by m1nsk** [comment]
Базу в нашем учебном случае легче сделать вектором, по нему поиск простой, просто указателям по полям скакать. Опять же нам по заданию нужно реализовать класс Ticket и дочерние, а не базу. Базу достаточно сделать статичной, только для тестов, без create/insert/delete. Нам по заданию даже кассу не нужно, я ее в плане поставил в пятом пункте. Да и вообще класс Ticket-абстрактный, в нем даже не должно быть конструктора, деструктора, я их написал только для теста. У нас же не может быть билета ни на что? ни на поезд ни на самолет а просто билет и все? Завтра хотя бы опишите дочерние классы билета, подводную лодку там, самолет. Вот у них и конструктор и деструктор обязаны быть.

"В общем согласен, что переключатель-автомат можно здесь грохнуть и просто продавать как пирожки, поле у билета менять. Но в реальности всегда есть переходы состояния и действия в каждом case разные." да там на каждое поле могут быть кейсы, не только на статус билета. Я не специалист, приведи пример и мы опишем кейсы для статуса.
